### PR TITLE
Solve issues with full send buffer

### PR DESCRIPTION
**Pull request by sander-visser**
Causing commands to be dropped + eventloop being blocked
Preventing file send to remote side

### CODE REVIEW

**Comment by sander-visser** [comment]
I have not yet been able to run unit tests (or update unit tests)

**Comment by cogu** [comment]
Good work. Need to make full review when I'n back home

**Comment by sander-visser** [comment]
If a rmf gets many writes from an apx-es node between apx_es_fileManager_run() calls, then im a bit worried that one will need a very large messageQueueBuf.

Maybe it should be possible to trigger early execution of apx_es_fileManager_run (via callback notifying of full queue for instance) if the messageQueue gets full but the transmitHandler is ready for new writes?

Iv added a commit to at least bundle sequential writes in an attempt to reduce the number of write notifications if the writes are triggered sequentially...

**Comment by sander-visser** [comment]
> 
> 
> Good work. Need to make full review when I'n back home

Made style updates to that first commit after review comment

**Comment by sander-visser** [comment]
Squashed the 17 commits to one.

For review it might be easier to track changes looking at individual commits at: https://github.com/sander-visser/c-apx/tree/apx-es-implement-pending-cmd-presquash

**Comment by sander-visser** [comment]
On an embedded device the pull req increases the need for flash by about 1kByte, But reduces the RAM need per filemanager by 1kB. It also allow for significant reductions in external send buffers as well as message queue size...

bild

File: apx/apx-es/inc/apx_es_fileManager.h [hunk 1]
```diff
@@ -59,18 +48,21 @@ typedef struct apx_es_fileManager_tag
...
    apx_transmitHandler_t transmitHandler;
-
-   apx_file_t *curFile; //weak pointer to last accessed file
-
-   bool pendingWrite;
-   bool pendingCmd;
-   bool dropMessage;
-
-   apx_es_command_t cmdInfo;
+   //buffer used to pack multipple messages prior to transmitHandler.send()
```
Comment (nilseri) [inline apx/apx-es/inc/apx_es_fileManager.h]: multipple -> multiple

Comment (nilseri) [inline apx/apx-es/inc/apx_es_fileManager.h]: multipple->multiple

File: apx/apx-es/inc/apx_es_fileManager.h [hunk 2]
```diff
@@ -87,6 +79,7 @@ int8_t apx_es_fileManager_create(apx_es_fileManager_t *self, uint8_t *messageQue
 void apx_es_fileManager_attachLocalFile(apx_es_fileManager_t *self, apx_file_t *localFile);
 void apx_es_fileManager_requestRemoteFile(apx_es_fileManager_t *self, apx_file_t *requestedFile);
 
+// Ensure it normally can provide a APX_ES_FILEMANAGER_MAX_CMD_BUF_SIZE / APX_ES_FILE_WRITE_MSG_FRAGMENTATION_THRESHOLD sized send buffer
```
Comment (nilseri) [inline apx/apx-es/inc/apx_es_fileManager.h]: Unclear what "it" is.
Suggestion: "Ensure that the handler can provide..."

File: apx/apx-es/inc/apx_es_fileManager_cfg.h [hunk 3]
```diff
@@ -11,7 +11,25 @@
 // CONSTANTS AND DATA TYPES
 //////////////////////////////////////////////////////////////////////////////
 
+#ifndef APX_ES_FILEMANAGER_MAX_NUM_REQUEST_FILES
 #define APX_ES_FILEMANAGER_MAX_NUM_REQUEST_FILES 10
+#endif
+
+#ifndef APX_ES_FILEMANAGER_MAX_CMD_BUF_SIZE
```
Comment (nilseri) [inline apx/apx-es/inc/apx_es_fileManager_cfg.h:19]: You mentioned that this was probably not needed anymore.

File: apx/apx-es/inc/apx_es_fileManager_cfg.h [hunk 4]
```diff
@@ -11,7 +11,25 @@
...
+#endif
+
+#ifndef APX_ES_FILE_WRITE_MSG_FRAGMENTATION_THRESHOLD
+// Fragmentation will not occur for smaller files or write notifications
+// Do not configure below RMF_MIN_MSG_LEN
+#define APX_ES_FILE_WRITE_MSG_FRAGMENTATION_THRESHOLD (RMF_HIGH_ADDRESS_SIZE + sizeof(uint64_t))
+#endif
+
+#ifndef APX_ES_FILEMANAGER_OPTIMIZE_WRITE_NOTIFICATIONS
+#define APX_ES_FILEMANAGER_OPTIMIZE_WRITE_NOTIFICATIONS 1
```
Comment (nilseri) [inline apx/apx-es/inc/apx_es_fileManager_cfg.h:33]: Add comment about what this define does.

File: apx/apx-es/test/testsuite_apx_es_filemanager.c [hunk 5]
```diff
@@ -18,7 +18,6 @@
 #define RECEIVE_BUFFER_LEN 4096
 #define APX_FILE_MANAGER_MSG_QUEUE_SIZE (sizeof(apx_msg_t)*APX_FILE_MANAGER_MAX_NUM_MESSAGES)
 //#define TRANSMIT_BUF_SIZE 4096
```
Comment (nilseri) [inline apx/apx-es/test/testsuite_apx_es_filemanager.c]: Remove commented out define?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 6]
```diff
@@ -1,17 +1,23 @@
 //////////////////////////////////////////////////////////////////////////////
-// INCLUDES
+// DEFINES
 //////////////////////////////////////////////////////////////////////////////
 #ifndef APX_DEBUG_ENABLE
 #define APX_DEBUG_ENABLE 0
 #endif
+#ifndef APX_MSQ_QUEUE_WARN_THRESHOLD
+#define APX_MSQ_QUEUE_WARN_THRESHOLD 2
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Add comment to explain what this define does?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 7]
```diff
@@ -202,8 +270,13 @@ void apx_es_fileManager_onFileUpdate(apx_es_fileManager_t *self, apx_file_t *fil
 void apx_es_fileManager_run(apx_es_fileManager_t *self)
 {
    int32_t result;
+   if ( (self == 0) || !self->isConnected )
+   {
+      return;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:306]: Misra forbids early return iirc?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 8]
```diff
@@ -212,72 +285,194 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
+   // Trigger refresh of the transmitBufLen before loop
+   self->transmitBuf = 0;
+
+   result = -1;
+   while ( (result != 0) &&
+           (self->pendingWrite == false) &&
+           (self->pendingMsg.msgType == RMF_CMD_INVALID_MSG) )
    {
-      while(true)
+      result = apx_es_fileManager_runEventLoop(self);
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:357]: Maybe use a different variable name than result? Since it's not obvious from the name what a function called "runEventLoop" returns.

File: apx/apx-es/src/apx_es_fileManager.c [hunk 9]
```diff
@@ -22,13 +28,24 @@
...
 static int32_t apx_es_fileManager_runEventLoop(apx_es_fileManager_t *self);
-static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self, apx_msg_t *msg);
+static int32_t apx_es_fileManager_processPendingMessage(apx_es_fileManager_t *self);
 static void apx_es_fileManager_parseCmdMsg(apx_es_fileManager_t *self, const uint8_t *msgBuf, int32_t msgLen);
 static void apx_es_fileManager_parseDataMsg(apx_es_fileManager_t *self, uint32_t address, const uint8_t *dataBuf, int32_t dataLen, bool more_bit);
 static void apx_es_fileManager_processRemoteFileInfo(apx_es_fileManager_t *self, const rmf_fileInfo_t *fileInfo);
 static void apx_es_fileManager_processOpenFile(apx_es_fileManager_t *self, const rmf_cmdOpenFile_t *cmdOpenFile);
 static int32_t apx_es_processPendingWrite(apx_es_fileManager_t *self);
-static int32_t apx_es_processPendingCmd(apx_es_fileManager_t *self);
+static inline int32_t apx_es_genFileSendMsg(uint8_t* msgBuf, uint32_t headerLen,
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:39]: Maybe add a comment for this function. For example what is "more_bit"? Could also name it "is_fragmented" / "more_fragments"?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 10]
```diff
@@ -186,13 +207,60 @@ void apx_es_fileManager_onMsgReceived(apx_es_fileManager_t *self, const uint8_t
...
       apx_msg_t msg = {RMF_MSG_WRITE_NOTIFY, 0, 0, 0}; //{msgType,msgData1,msgData2,msgData3}
       msg.msgData1=offset;
       msg.msgData2=length;
       msg.msgData3=(void*)file;
-      rbfs_insert(&self->messageQueue,(const uint8_t*) &msg);
+      if (self->queuedWriteNotify.msgType == RMF_MSG_WRITE_NOTIFY)
+      {
+         // Check if sequential write to the last file. If so append to the queued write notification
+         uint32_t potentialSequentialWriteSize = self->queuedWriteNotify.msgData2 + length;
+         if ( (msg.msgData3 == self->queuedWriteNotify.msgData3) &&
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Add a comment for "msgData3 == msgData3" (or checkSameFile function) - to explain what is wanted here

File: apx/apx-es/src/apx_es_fileManager.c [hunk 11]
```diff
@@ -186,13 +207,60 @@ void apx_es_fileManager_onMsgReceived(apx_es_fileManager_t *self, const uint8_t
...
+              ( (self->queuedWriteNotify.msgData1 + self->queuedWriteNotify.msgData2) == offset) &&
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: // check that the offset matches, i.e the write is sequential

File: apx/apx-es/src/apx_es_fileManager.c [hunk 12]
```diff
@@ -186,13 +207,60 @@ void apx_es_fileManager_onMsgReceived(apx_es_fileManager_t *self, const uint8_t
...
+            self->queuedWriteNotify.msgData2 = potentialSequentialWriteSize;
+         }
+#if APX_ES_FILEMANAGER_OPTIMIZE_WRITE_NOTIFICATIONS == 1
+         else if ( (self->queuedWriteNotify.msgData1 > offset) ||
+                   ( (self->queuedWriteNotify.msgData1 + self->queuedWriteNotify.msgData2) <
+                     (offset + length) ) )
+         {
+            // Not writing inside the queuedWriteNotify
+#else
+         else
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:258]: This is hard to follow, with different "else" following #else

File: apx/apx-es/src/apx_es_fileManager.c [hunk 13]
```diff
@@ -202,8 +270,13 @@ void apx_es_fileManager_onFileUpdate(apx_es_fileManager_t *self, apx_file_t *fil
...
 {
    int32_t result;
+   if ( (self == 0) || !self->isConnected )
+   {
+      return;
+   }
    if (self->pendingWrite == true)
    {
+      apx_es_setupTransmitBuf(self);
       result = apx_es_processPendingWrite(self);
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:311]: Maybe "continueOngoingWrite" or "writeNextFragment"?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 14]
```diff
@@ -212,72 +285,194 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
       {
 #if APX_DEBUG_ENABLE
-         fprintf(stderr, "apx_es_processPendingCmd returned %d",result);
+         fprintf(stderr, "pendingMsg processing returned %d",result);
 #endif
       }
+      else if (result > 0)
+      {
+         self->pendingMsg.msgType = RMF_CMD_INVALID_MSG;
+         self->transmitBufUsed = (uint32_t) result;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: transmitBufAmountUsed maybe? Or bytes used? only bufUsed sounds like a true/false.

File: apx/apx-es/src/apx_es_fileManager.c [hunk 15]
```diff
@@ -212,72 +285,194 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
    }
-   if ( (self->pendingWrite == false) && (self->pendingCmd == false) )
+
+   // Trigger refresh of the transmitBufLen before loop
+   self->transmitBuf = 0;
+
+   result = -1;
+   while ( (result != 0) &&
+           (self->pendingWrite == false) &&
+           (self->pendingMsg.msgType == RMF_CMD_INVALID_MSG) )
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Should be != RF_CMD_INVALID_MSG? Otherwise loop will break/not run if it is anything BUT invalid message?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 16]
```diff
@@ -212,72 +285,194 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
-         {
-            //MISRA
-         }
       }
+      else
+      {
+         // 0 = no work performed this loop
+      }
+   }
+   // Transmit pending data if any
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: // Transmit any remaining data (that was not flushed in the loop above)

File: apx/apx-es/src/apx_es_fileManager.c [hunk 17]
```diff
@@ -212,72 +285,194 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
+/**
+ * This is the apx_es_fileManager worker (by the main event loop)
+ * Prior to call the transmitBuf needs to be prepared
+ */
+static int32_t apx_es_fileManager_processPendingMessage(apx_es_fileManager_t *self)
+{
+   int32_t retval = 0;
+   if (self->transmitBuf == 0)
+   {
+      retval = -1;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Should this be an error? Or just a debugprint and return 0?

File: apx/apx-es/src/apx_es_fileManager.c [hunk 18]
```diff
@@ -286,176 +481,123 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
+            offset = self->pendingMsg.msgData1;
+            dataLen = self->pendingMsg.msgData2;
+            file = (apx_file_t*) self->pendingMsg.msgData3;
             //calculate address
             address = file->fileInfo.address+offset;
             headerLen = (address < RMF_DATA_HIGH_MIN_ADDR)? RMF_LOW_ADDRESS_SIZE : RMF_HIGH_ADDRESS_SIZE;
             msgLen = headerLen+dataLen;
-            sendBuffer = self->transmitHandler.getSendBuffer(self->transmitHandler.arg, msgLen);
-            if (sendBuffer != 0)
+            //Attempt to deliver the notification as one non-fragmented write
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Move comment inside the if stmt (as in the else below)

File: apx/apx-es/src/apx_es_fileManager.c [hunk 19]
```diff
@@ -286,176 +481,123 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
-               self->fileWriteInfo.readOffset=offset;
-               self->fileWriteInfo.writeAddress=address;
-               self->fileWriteInfo.remain=dataLen;
+               self->fileWriteInfo.localFile = file;
+               self->fileWriteInfo.readOffset = offset;
+               self->fileWriteInfo.writeAddress = address;
+               self->fileWriteInfo.remain = dataLen;
+               if (sendAvail >= (APX_ES_FILE_WRITE_MSG_FRAGMENTATION_THRESHOLD * 4u))
+               {
+                  // If it is a very large write get started despite fragmentation
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Update comment (has more to do with the size of sendAvail than the size of the write)

File: apx/apx-es/src/apx_es_fileManager.c [hunk 20]
```diff
@@ -286,176 +481,123 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
       case RMF_MSG_FILE_SEND: //sends file to remote side
          {
-            apx_file_t *file = (apx_file_t*) msg->msgData3;
-
-            if (sendAvail >= (int32_t) RMF_MIN_MSG_LEN)
+            apx_file_t *file = (apx_file_t*) self->pendingMsg.msgData3;
+            uint32_t dataLen = file->fileInfo.length;
+            uint32_t headerLen = (file->fileInfo.address < RMF_DATA_LOW_MAX_ADDR)? RMF_LOW_ADDRESS_SIZE : RMF_HIGH_ADDRESS_SIZE;
+            uint32_t msgLen = dataLen + headerLen;
+            bool more_bit = msgLen > sendAvail;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: rename more_bit

File: apx/apx-es/src/apx_es_fileManager.c [hunk 21]
```diff
@@ -466,14 +608,48 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
-   return -1; //invalid arguments
+   if (retval >= 0)
+   {
+      // Mark message as handled
+      self->pendingMsg.msgType = RMF_CMD_INVALID_MSG;
+   }
+   return retval;
+}
+
+static inline int32_t apx_es_genFileSendMsg(uint8_t* msgBuf, uint32_t headerLen,
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:659]: rename more_bit

File: apx/apx-es/src/apx_es_fileManager.c [hunk 22]
```diff
@@ -286,176 +481,123 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
+            //Attempt to deliver the notification as one non-fragmented write
+            if (msgLen <= sendAvail)
             {
-               int32_t result = rmf_packHeader(&sendBuffer[0], msgLen, address, false);
-               if (result > 0)
+               if (headerLen == rmf_packHeader(msgBuf, headerLen, address, false))
                {
-                  apx_file_read(file,&sendBuffer[headerLen],offset,dataLen);
-                  self->transmitHandler.send(self->transmitHandler.arg,0,msgLen);
+                  apx_file_read(file, &msgBuf[headerLen], offset, dataLen);
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: Handle apx_file_read() return value

File: apx/apx-es/src/apx_es_fileManager.c [hunk 23]
```diff
@@ -466,14 +608,48 @@ static int32_t apx_es_fileManager_onInternalMessage(apx_es_fileManager_t *self,
...
+   if (headerLen != rmf_packHeaderBeforeData(&msgBuf[headerLen], headerLen, file->fileInfo.address, more_bit))
+   {
+      retval = -1;
+   }
+   if (retval == 0)
+   {
+      retval = msgLen;
+      if (dataLen < file->fileInfo.length)
+      {
+         self->pendingWrite = true;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:677]: // Delay to next run cycle

File: apx/apx-es/src/apx_es_fileManager.c [hunk 24]
```diff
@@ -543,7 +719,7 @@ static void apx_es_fileManager_parseCmdMsg(apx_es_fileManager_t *self, const uin
  */
 static void apx_es_fileManager_parseDataMsg(apx_es_fileManager_t *self, uint32_t address, const uint8_t *dataBuf, int32_t dataLen, bool more_bit)
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c:760]: possible rename more_bit to be consistent

File: apx/apx-es/src/apx_es_fileManager.c [hunk 25]
```diff
@@ -562,13 +738,14 @@ static void apx_es_fileManager_parseDataMsg(apx_es_fileManager_t *self, uint32_t
                //start multi-message recepetion
                self->curFile = remoteFile;
                self->receiveStartAddress = address;
-               self->receiveBufOffset=0;
-               memcpy(&self->receiveBuf[self->receiveBufOffset], dataBuf, dataLen);
-               self->receiveBufOffset+=dataLen;
+               memcpy(&self->receiveBuf[0], dataBuf, dataLen);
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: remove & and [0] according to Jens

File: apx/apx-es/src/apx_es_fileManager.c [hunk 26]
```diff
@@ -577,20 +754,26 @@ static void apx_es_fileManager_parseDataMsg(apx_es_fileManager_t *self, uint32_t
       }
       else
       {
-         //continuation reception
-         offset = address-self->curFile->fileInfo.address;
+         if (self->dropMessage)
+         {
+            offset = self->receiveBufOffset;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: assert (receiveBufOffset==0)

File: apx/apx-es/src/apx_es_fileManager.c [hunk 27]
```diff
@@ -698,44 +886,42 @@ static void apx_es_fileManager_processOpenFile(apx_es_fileManager_t *self, const
...
-      {
-         sendAvail = (uint32_t )self->transmitHandler.getSendAvail(self->transmitHandler.arg);
-      }
-      if (sendAvail >= (int32_t) RMF_MIN_MSG_LEN)
+      int32_t retval = 0;
+      uint32_t sendAvail = self->transmitBufLen - self->transmitBufUsed;
+      uint32_t headerLen = (self->fileWriteInfo.writeAddress < RMF_DATA_LOW_MAX_ADDR)? RMF_LOW_ADDRESS_SIZE : RMF_HIGH_ADDRESS_SIZE;
+      uint32_t dataLen = self->fileWriteInfo.remain;
+      uint32_t msgLen = headerLen + self->fileWriteInfo.remain;
+      bool more_bit = msgLen > sendAvail;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: more_bit

File: apx/apx-es/src/apx_es_fileManager.c [hunk 28]
```diff
@@ -698,44 +886,42 @@ static void apx_es_fileManager_processOpenFile(apx_es_fileManager_t *self, const
...
-      int32_t retval=0;
-      uint32_t sendAvail = 0;
-      if ( (self->transmitHandler.getSendAvail != 0) && (self->transmitHandler.getSendBuffer != 0) )
-      {
-         sendAvail = (uint32_t )self->transmitHandler.getSendAvail(self->transmitHandler.arg);
-      }
-      if (sendAvail >= (int32_t) RMF_MIN_MSG_LEN)
+      int32_t retval = 0;
+      uint32_t sendAvail = self->transmitBufLen - self->transmitBufUsed;
+      uint32_t headerLen = (self->fileWriteInfo.writeAddress < RMF_DATA_LOW_MAX_ADDR)? RMF_LOW_ADDRESS_SIZE : RMF_HIGH_ADDRESS_SIZE;
```
Comment (nilseri) [inline apx/apx-es/src/apx_es_fileManager.c]: extract getHeaderLength function

Comment (sander-visser) [reply apx/apx-es/src/apx_es_fileManager.c]: done. found another bug in the process

File: apx/apx-es/src/apx_es_fileManager.c (hunk 15 above)
Comment (sander-visser) [reply apx/apx-es/src/apx_es_fileManager.c]: No, we want to run loop unti we encounter a msg that can not be serialized du to transmithandler buffer being full. Broke the check out to inline function to make it more clear...

File: apx/apx-es/inc/apx_es_fileManager_cfg.h [hunk 29]
```diff
@@ -11,7 +11,27 @@
...
 
+#ifndef APX_ES_FILEMANAGER_MAX_NUM_REQUEST_FILES
+// Number of remote files supported
 #define APX_ES_FILEMANAGER_MAX_NUM_REQUEST_FILES 10
+#endif
+
+#ifndef APX_ES_FILEMANAGER_MAX_CMD_BUF_SIZE
+// CMD_FILE_INFO_BASE_SIZE + RMF_HIGH_ADDRESS_SIZE + CMD_FILE_INFO_BASE_SIZE +
+// max supported file name len incl suffix and str termination
+#define APX_ES_FILEMANAGER_MAX_CMD_BUF_SIZE 256u
```
Comment (cogu) [inline apx/apx-es/inc/apx_es_fileManager_cfg.h:22]: This define is not used

File: apx/apx-es/src/apx_es_fileManager.c [hunk 30]
```diff
@@ -212,273 +316,385 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
 #endif
       }
    }
-   if (self->pendingCmd == true)
+   if (apx_es_isPendingMessage(&self->pendingMsg))
```
Comment (cogu) [inline apx/apx-es/src/apx_es_fileManager.c:319]: The entire body for apx_es_fileManager_run is now incredibly hard to read

File: apx/apx-es/src/apx_es_fileManager.c [hunk 31]
```diff
@@ -212,273 +316,385 @@ void apx_es_fileManager_run(apx_es_fileManager_t *self)
...
-            if (msgLen<=sendAvail)
+            cmdOpenFile.address = self->pendingMsg.msgData1;
+            if (msgLen <= sendAvail)
             {
-               uint8_t* msgBuf = self->transmitHandler.getSendBuffer(self->transmitHandler.arg, msgLen);
-               if (msgBuf != 0)
+               // Mark the remote file as open
+               apx_file_open((apx_file_t*) self->pendingMsg.msgData3);
+
+               retval = msgLen;
```
Comment (cogu) [inline apx/apx-es/src/apx_es_fileManager.c:530]: Don't you get signed/unsigned mismatch warning here?

**Review by cogu — APPROVED** [review]
